### PR TITLE
Format on open curly

### PR DESCRIPTION
**Pull request by aozgaa**
Fixes a vsfeedback issue: https://developercommunity.visualstudio.com/content/problem/11450/javascript-auto-format-of-function-curly-brace-not.html

Our brace-formatting behavior now follows that of CSharp more closely. Specifically, we now move opening curly braces onto a new line -- or move it back to the parent statement's line if it was on a new line -- under the following conditions:
* on typing an open curly
* on typing newline after a curly

We no longer require that the block statement be multi-line to remove newlines. Since VS performs brace-completions and inserts the closing brace on the same line, we need to adjust the formatting rules to be activated when there is a single-line block. We could alternatively make the rule check that the block is empty, but the change here is conceptually simpler.

### CODE REVIEW

File: src/services/formatting/formatting.ts [hunk 1]
```diff
@@ -97,11 +97,23 @@ namespace ts.formatting {
     }
 
     export function formatOnSemicolon(position: number, sourceFile: SourceFile, rulesProvider: RulesProvider, options: FormatCodeSettings): TextChange[] {
-        return formatOutermostParent(position, SyntaxKind.SemicolonToken, sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnSemicolon);
+        const outermostParent = findOutermostParentWithinListLevelFromPosition(position, SyntaxKind.SemicolonToken, sourceFile);
```
Comment (amcasey) [inline src/services/formatting/formatting.ts]: Checking for a semicolon token and formatting seem like separate operations.  Checking the token type here might simplify `findOutermostParentWithinListLevelFromPosition`.

Comment (amcasey) [inline src/services/formatting/formatting.ts]: Also, it seems like the thing we want to format will always be the immediate parent of the semicolon (modulo some padding to get full lines).

File: src/services/formatting/formatting.ts [hunk 2]
```diff
@@ -97,11 +97,23 @@ namespace ts.formatting {
     }
 
     export function formatOnSemicolon(position: number, sourceFile: SourceFile, rulesProvider: RulesProvider, options: FormatCodeSettings): TextChange[] {
-        return formatOutermostParent(position, SyntaxKind.SemicolonToken, sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnSemicolon);
+        const outermostParent = findOutermostParentWithinListLevelFromPosition(position, SyntaxKind.SemicolonToken, sourceFile);
+        return formatOutermostParent(outermostParent, sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnSemicolon);
```
Comment (amcasey) [inline src/services/formatting/formatting.ts]: It seems like this is just formatting a node.  Having "OutermostParent" in the name suggests that it might also navigate within the tree.

File: src/services/formatting/formatting.ts [hunk 3]
```diff
@@ -97,11 +97,23 @@ namespace ts.formatting {
...
     export function formatOnSemicolon(position: number, sourceFile: SourceFile, rulesProvider: RulesProvider, options: FormatCodeSettings): TextChange[] {
-        return formatOutermostParent(position, SyntaxKind.SemicolonToken, sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnSemicolon);
+        const outermostParent = findOutermostParentWithinListLevelFromPosition(position, SyntaxKind.SemicolonToken, sourceFile);
+        return formatOutermostParent(outermostParent, sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnSemicolon);
+    }
+
+    export function formatOnOpeningCurly(position: number, sourceFile: SourceFile, rulesProvider: RulesProvider, options: FormatCodeSettings): TextChange[] {
+        const openingCurly = findPrecedingTokenOfKind(position, SyntaxKind.FirstPunctuation, sourceFile);
+        const block = openingCurly && openingCurly.parent;
+        if (!(block && isBlock(block))) {
```
Comment (amcasey) [inline src/services/formatting/formatting.ts]: As we discussed offline, in bad code in might be the SourceFile.

File: src/services/formatting/formatting.ts [hunk 4]
```diff
@@ -121,44 +133,55 @@ namespace ts.formatting {
...
         const span = {
             pos: getLineStartPositionForPosition(parent.getStart(sourceFile), sourceFile),
             end: parent.end
         };
+
         return formatSpan(span, sourceFile, options, rulesProvider, requestKind);
     }
 
-    function findOutermostParent(position: number, expectedTokenKind: SyntaxKind, sourceFile: SourceFile): Node {
+    function findPrecedingTokenOfKind(position: number, expectedTokenKind: SyntaxKind, sourceFile: SourceFile): Node | undefined {
```
Comment (amcasey) [inline src/services/formatting/formatting.ts]: The name suggests that the method will walk backwards until it finds a token of the specified kind.  The actual behavior seems to be that it will find the unique token ending at the given position and return it if it is of the expected kind.

File: src/services/formatting/formattingContext.ts [hunk 5]
```diff
@@ -47,14 +47,6 @@ namespace ts.formatting {
             return this.contextNodeAllOnSameLine;
         }
 
-        public NextNodeAllOnSameLine(): boolean {
```
Comment (amcasey) [inline src/services/formatting/formattingContext.ts]: Why was this removed?

File: tests/cases/fourslash/formatOnEnterOpenBraceAddNewLine.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+/// <reference path="fourslash.ts"/>
+
+//// if(true) {/*0*/}
+//// if(false)/*1*/{
+//// }
+
+format.setOption("PlaceOpenBraceOnNewLineForControlBlocks", true);
+goTo.marker("0");
+edit.insertLine("");
+goTo.marker("1");
```
Comment (amcasey) [inline tests/cases/fourslash/formatOnEnterOpenBraceAddNewLine.ts]: Would it be interesting to check the file contents before inserting the second newline (to confirm that it's not formatting too much)?

File: tests/cases/fourslash/formattingOnSingleLineBlocks.ts [hunk 7]
```diff
@@ -1,16 +1,11 @@
 /// <reference path='fourslash.ts' />
 
-/////*1*/class C
-////{}/*2*/
-/////*3*/if (true)
-////{}/*4*/
+////class C
```
Comment (amcasey) [inline tests/cases/fourslash/formattingOnSingleLineBlocks.ts]: Would it be interesting to test the other places braces are allowed?  e.g. object literals, lambda bodies, etc?

File: src/services/formatting/formatting.ts (hunk 3 above)
Comment (aozgaa) [inline src/services/formatting/formatting.ts]: Checked the claim about open-curly parents. An opening curly can be typed as part of various constructs where there is no block in the AST (eg: object literals, type-literals, class declarations, interface declarations, binding names, etc). It seems reasonable to do formatting on typing the open curly in those cases.

So if we want to do formatting for the code following the curly, we should remove the `isBlock` check. Since we would format code *after* the open curly, and that may be surprising to some users, we should make the option toggle-able in some menu in VS.

File: src/services/formatting/formatting.ts (hunk 1 above)
Comment (aozgaa) [inline src/services/formatting/formatting.ts]: From out tests, consider
``` ts
if (true) {
if (true)
let x = 10
}
```
On typing a semi-colon after `10`, it's reasonable to format not only the statement, but the enclosing conditionals, yielding
``` ts
if (true) {
    if (true)
        let x = 10;
}
```

File: tests/cases/fourslash/formattingOnSingleLineBlocks.ts (hunk 7 above)
Comment (aozgaa) [inline tests/cases/fourslash/formattingOnSingleLineBlocks.ts]: Many constructs are already tested:
* object literals: `formattingOnObjectLiteral.ts` (see marker 26 for an empty object literal)
* type literals: `formatAnyTypeLiteral.ts`
* lambda: `formattingOnMultiLineBlockConstructs.ts` (marker 10 has an empty lambda).

We could reorganize these tests to be a bit more systematic, but it's unclear this would be worth the effort right now.

File: src/services/formatting/formatting.ts (hunk 1 above)
Comment (amcasey) [inline src/services/formatting/formatting.ts]: Wouldn't that entail formatting across multiple list levels?  I had the impression that we wanted to respect list boundaries.

File: src/services/formatting/formatting.ts [hunk 8]
```diff
@@ -121,49 +129,43 @@ namespace ts.formatting {
...
-        //
-        // let x = 1;
-        // while (true) {
-        // }
-        // after typing close curly in while statement we want to reformat just the while statement.
-        // However if we just walk upwards searching for the parent that has the same end value -
-        // we'll end up with the whole source file. isListElement allows to stop on the list element level
-        let current = precedingToken;
+    /**
+     * Finds and formats the highest node enclosing `position` whose end does not exceed the given position
```
Comment (amcasey) [inline src/services/formatting/formatting.ts]: What's "position"?

File: src/services/formatting/formatting.ts [hunk 9]
```diff
@@ -121,49 +129,43 @@ namespace ts.formatting {
...
-            return [];
-        }
-        const span = {
-            pos: getLineStartPositionForPosition(parent.getStart(sourceFile), sourceFile),
-            end: parent.end
-        };
-        return formatSpan(span, sourceFile, options, rulesProvider, requestKind);
-    }
+    /**
+     * Validating `expectedLastToken` ensures the token was typed in the context we expect (eg: not a comment).
```
Comment (amcasey) [inline src/services/formatting/formatting.ts]: Comment out of date.

File: src/services/formatting/formatting.ts [hunk 10]
```diff
@@ -97,11 +97,29 @@ namespace ts.formatting {
...
     export function formatOnSemicolon(position: number, sourceFile: SourceFile, rulesProvider: RulesProvider, options: FormatCodeSettings): TextChange[] {
-        return formatOutermostParent(position, SyntaxKind.SemicolonToken, sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnSemicolon);
+        const semicolon = findImmediatelyPrecedingTokenOfKind(position, SyntaxKind.SemicolonToken, sourceFile);
+        return formatNodeLines(findOutermostNodeWithinListLevel(semicolon), sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnSemicolon);
+    }
+
+    export function formatOnOpeningCurly(position: number, sourceFile: SourceFile, rulesProvider: RulesProvider, options: FormatCodeSettings): TextChange[] {
+        const openingCurly = findImmediatelyPrecedingTokenOfKind(position, SyntaxKind.OpenBraceToken, sourceFile);
+        if (openingCurly) {
+            const curlyBraceRange = openingCurly.parent;
```
Comment (amcasey) [inline src/services/formatting/formatting.ts]: Consider adding a comment here about unbalanced braces and disabled Brace Completion.

File: src/services/formatting/formatting.ts [hunk 11]
```diff
@@ -97,11 +97,29 @@ namespace ts.formatting {
...
 
     export function formatOnSemicolon(position: number, sourceFile: SourceFile, rulesProvider: RulesProvider, options: FormatCodeSettings): TextChange[] {
-        return formatOutermostParent(position, SyntaxKind.SemicolonToken, sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnSemicolon);
+        const semicolon = findImmediatelyPrecedingTokenOfKind(position, SyntaxKind.SemicolonToken, sourceFile);
+        return formatNodeLines(findOutermostNodeWithinListLevel(semicolon), sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnSemicolon);
+    }
+
+    export function formatOnOpeningCurly(position: number, sourceFile: SourceFile, rulesProvider: RulesProvider, options: FormatCodeSettings): TextChange[] {
+        const openingCurly = findImmediatelyPrecedingTokenOfKind(position, SyntaxKind.OpenBraceToken, sourceFile);
+        if (openingCurly) {
```
Comment (amcasey) [inline src/services/formatting/formatting.ts]: Consider inverting this `if` to reduce rightward drift.

File: src/services/formatting/formatting.ts [hunk 12]
```diff
@@ -97,23 +97,19 @@ namespace ts.formatting {
...
     export function formatOnOpeningCurly(position: number, sourceFile: SourceFile, rulesProvider: RulesProvider, options: FormatCodeSettings): TextChange[] {
-        const openingCurly = findPrecedingTokenOfKind(position, SyntaxKind.FirstPunctuation, sourceFile);
-        const block = openingCurly && openingCurly.parent;
-        if (!(block && isBlock(block))) {
-            return [];
-        }
-        const outermostParent = findOutermostParentWithinListLevel(block);
-        return formatOutermostParent(outermostParent, sourceFile, options, rulesProvider, FormattingRequestKind.FormatOnOpeningCurlyBrace);
+        const openingCurly = findImmediatelyPrecedingTokenOfKind(position, SyntaxKind.OpenBraceToken, sourceFile);
+        const curlyBraceRange = openingCurly && openingCurly.parent;
```
Comment (amcasey) [inline src/services/formatting/formatting.ts]: Explained offline that the matching-endpoint computation depends on it.

**Comment by amcasey** [comment]
Thanks for making those changes!

**Comment by andy-ms** [comment]
> Our brace-formatting behavior now follows that of CSharp more closely. Specifically, we now move opening curly braces onto a new line ... on typing an open curly

Won't this be very annoying in our own codebase?

**Comment by CyrusNajmabadi** [comment]
> Our brace-formatting behavior now follows that of CSharp more closely. Specifically, we now move opening curly braces onto a new line ... on typing an open curly

I don't think C# does that...

openbrace

**Comment by aozgaa** [comment]
@CyrusNajmabadi I miss-spoke, we only move the open-curly to a new-line when the block becomes a multi-line block. The description has been updated. Sorry for the confusion.

PS: how are you recording your typing?

@andy-ms See above, and moreover, this only occurs if the user has set the brace-style through the formatting settings. So we only make changes that make the code more-resemble the formatting settings they are using. The defaults for the brace-on-newline settings are false; braces are on the same-line by default.

Finally, this fixes a regression in this functionality in VS reported in the feedback issue above.

**Comment by CyrusNajmabadi** [comment]
@aozgaa I use "screentogif".  Fantastic program for this sort of thing.
